quesma/quesma/functionality/bulk/bulk.go

```diff
@@ -3,110 +3,260 @@
 
-		switch operation {
-		case "create", "index":
-			results = append(results, WriteResult{operation, index})
-			indicesWithDocumentsToInsert[index] = append(indicesWithDocumentsToInsert[index], document)
-		case "update":
-
-			errorstats.GlobalErrorStatistics.RecordKnownError("_bulk update is not supported", nil,
-				"We do not support 'update' in _bulk")
-			logger.Debug().Msg("Not supporting 'update' _bulk.")
-
-		case "delete":
-			errorstats.GlobalErrorStatistics.RecordKnownError("_bulk delete is not supported", nil,
-				"We do not support 'delete' in _bulk")
-			logger.Debug().Msg("Not supporting 'delete' _bulk.")
-
-		default:
-			errorstats.GlobalErrorStatistics.RecordUnknownError(nil,
-				fmt.Sprintf("Unexpected operation in _bulk: %v", operation))
-			logger.Error().Msgf("Invalid JSON with operation definition in _bulk: %s", operation)
-		}
+			documentBytes, err := document.Bytes()
+			if err != nil {
+				return err
+			}
+			elasticRequestBody = append(elasticRequestBody, documentBytes...)
+			elasticRequestBody = append(elasticRequestBody, '\n')
+
+			elasticBulkEntries = append(elasticBulkEntries, entryWithResponse)
+		} else {
+			// Bulk entry for Clickhouse
+			if operation != "create" && operation != "index" {
+				// Elastic also fails the entire bulk in such case
+				logger.ErrorWithCtxAndReason(ctx, "unsupported bulk operation type").Msgf("unsupported bulk operation type: %s", operation)
+				return fmt.Errorf("unsupported bulk operation type: %s. Operation: %v, Document: %v", operation, rawOp, document)
+			}
 
+			clickhouseDocumentsToInsert[index] = append(clickhouseDocumentsToInsert[index], entryWithResponse)
+		}
+		return nil
 	})
 
+	return results, clickhouseDocumentsToInsert, elasticRequestBody, elasticBulkEntries, err
+}
+
+func sendToElastic(elasticRequestBody []byte, cfg config.QuesmaConfiguration, elasticBulkEntries []BulkRequestEntry) error {
+	if len(elasticRequestBody) == 0 {
+		// Fast path - no need to contact Elastic!
+		return nil
+	}
+
+	req, _ := http.NewRequest("POST", cfg.Elasticsearch.Url.String()+"/_bulk", bytes.NewBuffer(elasticRequestBody))
+	req.Header.Set("Content-Type", "application/x-ndjson")
+	client := http.Client{} // FIXME
+	response, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+
+	if response.StatusCode != http.StatusOK {
+		responseBody, _ := io.ReadAll(response.Body)
+		return fmt.Errorf("error sending bulk request (%v): %s", response.StatusCode, responseBody)
+	}
+
+	responseBody, err := io.ReadAll(response.Body)
 	if err != nil {
-		logger.ErrorWithCtx(ctx).Msgf("Error processing _bulk: %v", err)
-		return
+		return err
 	}
 
-	for indexName, documents := range indicesWithDocumentsToInsert {
+	var elasticBulkResponse BulkResponse
+	err = json.Unmarshal(responseBody, &elasticBulkResponse)
+	if err != nil {
+		return err
+	}
+
+	// Copy Elastic's response entries to our response
+	for i, entry := range elasticBulkResponse.Items {
```

Comment:
Is result of this copy used somewhere after `sendToElastic` function call?

Reply:
`.response` that I write to is a pointer into `results` and we return `results` in `Write` function.

Generally (maybe I should add some comment) the whole thing works like this: We receive a bulk request that writes to multiple indexes, we group those bulk entries by index. However, by doing this grouping we are losing the order - 1st bulk entry should correspond to 1st response entry, 2nd bulk entry to 2nd response entry and so on. To combat that, I create a `results` array and each `BulkRequestEntry` has a pointer to the correct entry in the resulting array.

Reply:
I think comment would be useful

Reply:
Added comment in the last commit.